## Making ranges

In [3]:
type(range(15))

range

In [4]:
x = list(range(15))
x

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14]

In [5]:
L = [3,5,7,8]
for ind in range(len(L)):
    print(ind)

0
1
2
3


In [6]:


for ind in range(len(L)):
    print(L[ind])

3
5
7
8


In [7]:
list(range(1,5))

[1, 2, 3, 4]

In [8]:
list(range(5,100))

[5,
 6,
 7,
 8,
 9,
 10,
 11,
 12,
 13,
 14,
 15,
 16,
 17,
 18,
 19,
 20,
 21,
 22,
 23,
 24,
 25,
 26,
 27,
 28,
 29,
 30,
 31,
 32,
 33,
 34,
 35,
 36,
 37,
 38,
 39,
 40,
 41,
 42,
 43,
 44,
 45,
 46,
 47,
 48,
 49,
 50,
 51,
 52,
 53,
 54,
 55,
 56,
 57,
 58,
 59,
 60,
 61,
 62,
 63,
 64,
 65,
 66,
 67,
 68,
 69,
 70,
 71,
 72,
 73,
 74,
 75,
 76,
 77,
 78,
 79,
 80,
 81,
 82,
 83,
 84,
 85,
 86,
 87,
 88,
 89,
 90,
 91,
 92,
 93,
 94,
 95,
 96,
 97,
 98,
 99]

In [9]:
list(range(5,100,5))

[5, 10, 15, 20, 25, 30, 35, 40, 45, 50, 55, 60, 65, 70, 75, 80, 85, 90, 95]

In [10]:
list(range(10,0,-1))

[10, 9, 8, 7, 6, 5, 4, 3, 2, 1]

In [11]:
list(range(10,-1,-1))

[10, 9, 8, 7, 6, 5, 4, 3, 2, 1, 0]

In [12]:
type(x)

list

In [13]:
L = list(range(10,-1,-1))
print(L)
L[2:8]

[10, 9, 8, 7, 6, 5, 4, 3, 2, 1, 0]


[8, 7, 6, 5, 4, 3]

In [14]:
L[8:2:-1]

[2, 3, 4, 5, 6, 7]

In [15]:
L

[10, 9, 8, 7, 6, 5, 4, 3, 2, 1, 0]

In [16]:
L[::-1]

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

## 2.  Numpy arrays

In [17]:
import numpy as np
a = np.arange(15)
print(a)
a

[ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14]


array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14])

This is called an **array** and for now you can think of this as sequence with only one type.  You can do various sequence like things, with arrays, such as indexing by number and taking splices.

But we haven't yet seen what arrays really can do.   In addition to these sequence-like arrays there are also **2D arrays**, which are **tables of numbers**.  Here both row and column structure matter.

In [18]:
LL_list = list(range(15))
print(LL_list)
LL = np.array([LL_list[0:5],LL_list[5:10],LL_list[10:15]])
print(LL)

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14]
[[ 0  1  2  3  4]
 [ 5  6  7  8  9]
 [10 11 12 13 14]]


In [19]:
a[13]

13

In [20]:
LL[2,3]

13

In [21]:
LL[2,:]

array([10, 11, 12, 13, 14])

In [22]:
LL[2,0:2]

array([10, 11])

In [23]:
print(a)
a[2]

[ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14]


2

Here is a simpler way to define LL.  Use `np.arange`, which is like the default Python `arange`, except that it produces an array.  More on `reshape` below.

In [24]:
LL2 = np.arange(15).reshape((3,5))
print(LL2)

[[ 0  1  2  3  4]
 [ 5  6  7  8  9]
 [10 11 12 13 14]]


`a_1d` is a 1D array.  `a` is a 2D array, meaning it has rows and columns.

In [25]:
a_1d = np.arange(15)
print(a_1d)
a = a_1d.reshape(3, 5)
print(a)
b = np.arange(15).reshape(5, 3)
a = np.arange(15).reshape(3, 5)
print(b)
print(b.transpose())

[ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14]
[[ 0  1  2  3  4]
 [ 5  6  7  8  9]
 [10 11 12 13 14]]
[[ 0  1  2]
 [ 3  4  5]
 [ 6  7  8]
 [ 9 10 11]
 [12 13 14]]
[[ 0  3  6  9 12]
 [ 1  4  7 10 13]
 [ 2  5  8 11 14]]


The need for 2D arrays is obvious if you've taken a linear algebra class.  They correspond to the mathematical object called a **matrix**.  One application for matrices is in solving systems of equations, but this really only scratches the surface.  A more fundamental application for a data scientist is they can represent all that is known about a single data set.  Generally each row represents an item (an individual or event in the data), and the entry in each column its value for a particular attribute. For example let's say each room in a hotel has lamps, tables, chairs, and beds, but in varying numbers.  We might represent the inventory of items in 5-room hotel with a 5x4 matrix (5 rows, 4 columns), as follows:

```
6  3 4 1
5  2 3 2
8  3 6 2
5  1 3 1
10 4 7 2
```

So the first row represents a room with 6 lamps, 3 tables, 4 chairs, and 1 bed.
Now if we represent the cost of each item as 1D cost array (or **vector**) to use the mathematical term,

```
40 175 90 450,
```

where the costs are ordered in the same way as our columns above: lamps costs, table costs, chair costs, and bed costs.  Then we can compute the per room cost or the first room as follows:

In [26]:
6*40 + 3*175 + 4*90 + 1*450

1575

Now the computation above, the cost of the items in the first room, can also be done as the "dot product" (or "dot") of the first row of the room_matrix and the cost vector.  The dot product of two 1D arrays is just the sum of the product of the corresponding terms in the two arrays (they need to be the same length).  That is, 

```
6*40 + 3*175 + 4*90 + 1*450
```

In [27]:
room_matrix = \
np.array(
[[6,  3, 4, 1],
[5,  2, 3, 2],
[8,  3, 6, 2],
[5,  1, 3, 1],
[10, 4, 7, 2]])

cost_vector = np.array([40, 175, 90, 450])

In [28]:
cost_vector.dot(room_matrix[0,:])

1575

If we ask for the dot product of an M x N array A with a 1D N-array B, the result is a 1D array containing the dot product of the M rows of A with B.  Applied to our example, the dot product of the room matrix with the cost vector yields the costs of the 5 rooms.

In [29]:
print(room_matrix)
print(cost_vector)
print(room_matrix.dot(cost_vector))

[[ 6  3  4  1]
 [ 5  2  3  2]
 [ 8  3  6  2]
 [ 5  1  3  1]
 [10  4  7  2]]
[ 40 175  90 450]
[1575 1720 2285 1095 2630]


The mathematical name for the `dot` method, which computes the per room cost here, is **matrix multiplication**.

Arrays have many attributes.

In [30]:
print(a.shape)
print(a.ndim)          # a is a 2D arrayam
print(a.dtype.name)    # np.int64 maxint = 2**63 - 1
print(a.itemsize)      # Such an int takes up 8 bytes of memory
print(a.size)          # Number of elements
print(type(a))
print(np.ndarray)

(3, 5)
2
int64
8
15
<class 'numpy.ndarray'>
<class 'numpy.ndarray'>


The python interpreter and the print function will print out arrays differently (Although this is an option any Python type can take, most types look the same whether they're printed by `print` or the interpreter).  Either way, arrays look different from lists.

In [31]:
b = np.array([6, 7, 8])
print(b)
print(type(b))
b

[6 7 8]
<class 'numpy.ndarray'>


array([6, 7, 8])

A sequence of sequences can be used to define an array, but since the inner sequences are rows, they must all be the same length.

In [32]:
b = np.array( [ (1.5,2,3), (4,5,6) ] )
b

array([[1.5, 2. , 3. ],
       [4. , 5. , 6. ]])

Arrays can also contain complex numbers, but remember it takes two quantities to define a single complex number::

In [33]:
c = np.array( [ [1,2], [3,4] ], dtype=complex )
c

array([[1.+0.j, 2.+0.j],
       [3.+0.j, 4.+0.j]])

In [34]:
c = np.array( [ [1+1j,2-3j], [3+5j,4+1j] ], dtype=complex )
c

array([[1.+1.j, 2.-3.j],
       [3.+5.j, 4.+1.j]])

A very convenient way to fill an array is to start with an array containing all 0's or 1's and then update the contents:

In [35]:
X = np.zeros( (3,4) )
X

array([[0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.]])

## Reshaping of an array 

```
A.reshape((m,n))
```

means recasting the data in and `A` into
m rows and n columns.,  This means `A` has to have m x n 
cells.  Reshaping can be done on any array that has the right number
of cells.  For example, a 2D 3x4 array can be reshaped into a 6x2
array.

In [36]:
Y = np.arange(12).reshape((3,4))
print(Y)
Y[2,3]

[[ 0  1  2  3]
 [ 4  5  6  7]
 [ 8  9 10 11]]


11

## Indexing

Indexing works as it does with ordinary Python sequences, except that now there are two dimensions to worry about

In [37]:
L= list(range(12))
print(L)
L[0:3] = [3,4]
print(L)

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]
[3, 4, 3, 4, 5, 6, 7, 8, 9, 10, 11]


In [38]:
X = np.zeros((3,4))

In [39]:
X[0,0]

0.0

In [40]:
X[1,2]

0.0

In [41]:
print(X)
X[2,0]

[[0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]]


0.0

In [54]:
Y = np.arange(12).reshape((3,4))
print(Y)
print(Y[2,3])
print(Y[3,2])

[[ 0  1  2  3]
 [ 4  5  6  7]
 [ 8  9 10 11]]
11


IndexError: ignored

Reshaping an array creates a new array.  The original array can only be indexed according to its original shape:

In [55]:
Z = Y.reshape((4,3))
print(Z[3,2])
print(Y[3,2])

11


IndexError: ignored

However, the two arrays `Z` and `Y` still share the same **data**.  What that means is that changing one also changes the other:

In [56]:
print('Y:')
print(Y)
print('Z:')
print(Z)
print('    ==>')
Y[1,0] = 14
print('Y:')
print(Y)
print('Z:')
print(Z)

Y:
[[ 0  1  2  3]
 [ 4  5  6  7]
 [ 8  9 10 11]]
Z:
[[ 0  1  2]
 [ 3  4  5]
 [ 6  7  8]
 [ 9 10 11]]
    ==>
Y:
[[ 0  1  2  3]
 [14  5  6  7]
 [ 8  9 10 11]]
Z:
[[ 0  1  2]
 [ 3 14  5]
 [ 6  7  8]
 [ 9 10 11]]


Using `a` and `LL` as defined above:

In [ ]:
print(a)
a[13]

[ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14]


13

In [ ]:
print(LL)
LL[2,3]

[[ 0  1  2  3  4]
 [ 5  6  7  8  9]
 [10 11 12 13 14]]


13

An expression that returns the third row of `LL` as a 1D array:

In [ ]:
LL[2,:]

array([10, 11, 12, 13, 14])

An expression that returns a splice (the first two items) of the third row of `LL` as a 1D array:

In [ ]:
LL[2,0:2]

array([10, 11])

## Assignment

Assigment works as it does with Python sequences, except that it now works in two dimensions

In [ ]:
X[1,2] = 3
X

array([[0., 0., 0., 0.],
       [0., 0., 3., 0.],
       [0., 0., 0., 0.]])

In [ ]:
X = np.zeros( (3,4), dtype=int )
ctr = 0
(rows,cols) = X.shape
for i in range(rows):
    for j in range(cols):
        ctr += 1
        X[i,j] += ctr*10
X

array([[ 10,  20,  30,  40],
       [ 50,  60,  70,  80],
       [ 90, 100, 110, 120]])

In [ ]:
type(X[1,2])

numpy.int64

In [ ]:
import numpy as np
X = np.arange(10,130,10).reshape((3,4))
X

array([[ 10,  20,  30,  40],
       [ 50,  60,  70,  80],
       [ 90, 100, 110, 120]])

In [ ]:

print(X)
print(X[2,3])
print(X[2,:])
print(X[:,2])
print(X[1:3,1:3])

[[ 10  20  30  40  50]
 [ 60  70  80  90 100]
 [110 120 130 140 150]]
140
[110 120 130 140 150]
[ 30  80 130]
[[ 70  80]
 [120 130]]


In [ ]:
Y = np.arange(10,160,10).reshape((3,5))

The following expression gives the second row of `Y`

```
Y[1,:]
```

The following expression gives the 3rd column of `Y`:

```
Y[:,2]
```

Note that columns and rows are just 1D arrays; nothing identifies them as row vectors or column vectors.

As a convenience, `numpy` lets you skip the `:` when you want a row.  In other words, the next two expressions are synonyms.  Note that you **need** the `:` if you want a column.

In [ ]:
print(Y[1,:])
print(Y[1])

[4 5 6 7]
[4 5 6 7]


In [ ]:
print(X)
print(X[1,:], X[1,:].shape)
print(X[:,1])

[[ 10  20  30  40  50]
 [ 60  70  80  90 100]
 [110 120 130 140 150]]
[ 60  70  80  90 100] (5,)
[ 20  70 120]


Next let's get part of a column or row.  The first two elements of the second column of `X`, followed by the first two elements of the second row of `X`:

In [ ]:
print(X)
print(X[1,:2])
print(X[:2,1])

[[ 10  20  30  40  50]
 [ 60  70  80  90 100]
 [110 120 130 140 150]]
[60 70]
[20 70]


Now let's get a 2x2 subarray, consisting of the upper left hand corner of `X`, followed by the central part of the last two rows.

In [ ]:
print(X)
print(X[:2,:2])
print(X[1:,1:3])

[[ 10  20  30  40  50]
 [ 60  70  80  90 100]
 [110 120 130 140 150]]
[[10 20]
 [60 70]]
[[ 70  80]
 [120 130]]


Array transposition is sometimes a nice way to get to the 2D array you eally want.  The transposition of an array `M` is called `M.T`, and the definition is that 

```
M.T[i,j] = M[j,i]
```

So if `M` is an `m` x `n` array, then `M.T` is an `n` x `m`  array.   Look at `X.T` and verify these definitions.  The `m`th row of `X` becomes the `m`th column of `X.T`. The `n`th column  of `X` becomes the `n`th row of `X.T`.

In [ ]:
print(X)
print()
print(X.T)
print(X[1,2], X.T[2,1])

[[ 10  20  30  40  50]
 [ 60  70  80  90 100]
 [110 120 130 140 150]]

[[ 10  60 110]
 [ 20  70 120]
 [ 30  80 130]
 [ 40  90 140]
 [ 50 100 150]]
80 80


Another set of examples.  The point to observe here is that `a` is a 1-dimensional array, which means that only one index is required, or allowed!

In [ ]:
a = np.arange(10)**3
a

array([  0,   1,   8,  27,  64, 125, 216, 343, 512, 729])

What to do to see that it's 1-dimensional

In [ ]:
a.shape

(10,)

In [ ]:
a[2]

8

Attempting to access a 1D array with two indices is an `IndexError`:

In [ ]:
a[2,0]

IndexError: ignored

So as a first approximation,indexing arrays works much as indexing lists does.  But there are some important differences.  First arrays can have more than one dimension,. hence more than index.  Second, assignment works a little differently.

In [ ]:
print(a)
a[2:5]

[-1000     1 -1000     3 -1000     5     6     7     8     9    10    11
    12    13    14]


array([-1000,     3, -1000])

A big difference between an array and a list is that arrays are more flexible about assignments.  The value assigned to a splice can be a scalar, in which case the assignment is "broadcast" to each of the cells in the slice (we discuss broadcasting in more detail in another `numpy` notebook).

In th next cell we demonstrate assignment to a splice.  In normal Python the value that is assigned to a splice has to be a sequence.  But in numpy it is treated as an elementwise operation, so in the next  example, we assign the value -1000 to each of the positions in the splice.

In [ ]:
print(a)
a[:6:2] = -1000    # equivalent to a[0:6:2] = -1000; from start to position 6, exclusive, set every 2nd element to -1000
print(a)

[ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14]
[-1000     1 -1000     3 -1000     5     6     7     8     9    10    11
    12    13    14]


## 3 (and more) D arrays

There are also 3D arrays, which have a third dimension; each position along the third dimension defines a 2D array.

In [ ]:
A =np.arange(24).reshape((2,3,4))
A

 In a 2D array, a single index slice is a row or column (a 1D array).  In a 3D array, a single index slice is a 2D array.

In [ ]:
A[1,:,:]

In [ ]:
A[:,:,3]

In [ ]:
A[:,2,:]

We will not be using 3D arrays much, and they will not be featured on quizzes.  But it's nice to know they're out there. 

An easy example of where you might want a 3D array is a color image. Specifying each pixel position in the image takes two numbers, and in a black and white image we can just have the cell at that position contain a single number representing the grayscale magnitude.  But for a color image we need three numbers, and it is very convenient for many purposes to use a third dimension for those three numbers; for example, using the third dimension for the color, and the first "layer" in the 3rd dimension for red, there is an easily retrievable 2D array representing all the "red" values for the image `I`:

```
I[:,:, 0]
```

In machine learning applications, especially in deep learning applications, 3D arrays and higher are not at all uncommon.  One reason is that it is very convenient (and efficient) to use one dimension for the batch number.  Another is that words in the input are often represented as "vectors" (1D arrays of floating point numbers).  Then

```
D[22, 12,:]
```

retrieves the word vector for the 13th word in the 23rd batch.

## 3.  Elementwise arithmetic operations

In [ ]:
import numpy as np
a = np.array( [20,30,40,50] ).reshape((2,2))
b = np.arange( 4 ).reshape((2,2))
print(a)
print(b)

[[20 30]
 [40 50]]
[[0 1]
 [2 3]]


In [ ]:
a + b

array([[20, 31],
       [42, 53]])

In [ ]:
a * b

array([[  0,  30],
       [ 80, 150]])

In [ ]:
a - b

array([[20, 29],
       [38, 47]])

In the next example we briefly illustrate **broadcasting**.  Elementwise operations are extended to apply between arrays and objects that are not arrays.  So for example, 2 * a (a is the array above) returns an array that contains all the elements of a multiplied by 2

In [ ]:
print(a)
print()
print(2 * a)

[[20 30]
 [40 50]]

[[ 40  60]
 [ 80 100]]


This works by "broadcasting" 2 into an array the same size as a and then doing elementwise mutliplication on the two arrays.  We will say more about broadcasting in a subsequent notebook.

The next example applies two operations to `a`, first `sin`, then multiplication by 10.

You can multiply two arrays of the same shape together as we did with `a` and `b`
above, but you can't rely on broadcasting to figure out what to do with mismatched array sizes. In general, multiplying arrays of different sizes together will fail:

In [ ]:
c = np.array([3,2,1,5,4])
print(a)
c * a

The next example demonstrates the possibility of applying and broadcasting two operations to `a`, first `sin`, then multiplication by 10.

In [ ]:
10*np.sin(a)

The same thing works with Boolean tests, a point we will will return to below.

In [ ]:
print(a)
Y = a < 35
print(Y)

[[20 30]
 [40 50]]
[[ True  True]
 [False False]]


In [ ]:
a[Y]

In [ ]:
print(a)
print(b)
a == b

In [ ]:
if a == b:
    print("Hi!")

In [ ]:
A = np.array( [[1,1],
            [0,1]] )
B = np.array( [[2,0],
            [3,4]] )

In [ ]:
A * B  # elementwise product

## Fast and slow array creation

Here are two methods of filling in an array, one with a loop, another with the `numpy` builtin `np.fromfunction`, which needs a two argument function that will return the desired value for cell `[i,j]` in the array whden given the artgument `(i, j)`.

In [63]:
import numpy as np
import time

num_iters = 1000
(rows, cols) = (768,1024)

# Method A

start = time.time()
for iter_num in range(num_iters):
  X = np.zeros( (rows,cols), dtype=int )
  (rows,cols) = X.shape
  for i in range(rows):
      for j in range(cols):
          #X[i,j] = 10 * i + j
          X[i,j] = f(i,j)

print('A:', get_avg_duration(start, num_iters), 'secs/iter')


# Method B

# make a rowsxcols array  `b` using the function `f`
# where b[i,j] = f(i,j)
start = time.time()
for iter_num in range(num_iters):
    b = np.fromfunction(f,(rows,cols))
print('B:', get_avg_duration(start, num_iters), 'secs/iter')

A: 0.274717577457428 secs/iter
B: 0.0034362633228302002 secs/iter


The result on my machine was that using `np.fromfunction`  is a little under 2 orders of magnitude faster.  Try it on yours.

Note a key point here.  The differences between these two ways of array filling are dependent on the sizes of the arrays being filled.  The same code run with small (for example, 4x5) arrays is a virtual tie.

## Elementwise Boolean operations, Masking

A very important extension of elementwise operations is the extension to Boolean tests.

Applying a Boolean test to an array returns an array of truth-values: Just as

```
3 + X
```

adds 3 to every element of array `X`, so

```
X > 2
```

returns an array of truth-values which tells us which elements of `X` are greater than 2.

In [ ]:
import numpy as np
X = np.array([1,5,2,7,9,4,3,-6])
print(X)
print(X >= 3)

What is the point of having such things around?  

Here it is:  Boolean arrays can be used to index other arrays.  They must be the same shape as the array
they index.   So in the simplest case:

In [ ]:
print(X)
print(X[np.array([False,  True, False,  True,  True,  True,  True, False])])

But now I can take advantage of the fact `X>=3` is a Boolean array to 

In [ ]:
print(X)
print(X>=3)
print(X[X>=3])

This returns exactly the members of X that are greater than or equal to 3.

We could do the same with a list comprehension, but the array computation above is much faster:

In [ ]:
np.array([x for x in X if x >= 3])

Boolean arrays can also be used to **count** the number of elements in an array that satisfy some constraint. Although `len` works on arrays and could be used, `sum` is faster.  The Boolean `True` is treated as 1 and the Boolean `False` as 0, so summing a Boolean array counts
the number of trues.  The following expression correctly counts the number of elements in `X` that are greater than or equal to 3.

In [ ]:
sum(X>=3)

array([4, 4, 4, 5])

More on array operations and efficiency below.  It's a tricky subject.

The reason that `X[X>=3]` works has to do with a basic fact of array indexing we haven't
really made explicit yet.  Any Boolean array Y of the right length can be used to index an array X if X and Y are the same length.  

A hokey but essential example.  Suppose we have an array of length 8 and we want to access the first, third, and seventh elements.  We can of course do it with a Boolean array:

In [ ]:
X = np.arange(8) + 1
Y = np.array([True,False,True,False,False,False,True,False])
print(X)
X[Y]

But there's a better way!

Cook up a sequence consisting of the **indices** of the first, third, and seventh elements, namely:

```
[0,2,6]
```

Now use that sequence as a **fancy** index on X:

In [ ]:
X[[0,2,6]]

Note the difference between what we just did and passing three different indices to the 1D array `X`,
which is an error.

In [ ]:
X[0,2,6]

Fancy indexing also works on 2D arrays.

In [ ]:
Y = np.arange(12).reshape((3,4))
print(Y)
print(Y[[0,2],[1,2]])

Back to Boolean operations.  Boolean operations work more or less as expected on 2D arrays:

In [ ]:
import numpy as np
y = np.arange(35).reshape(5,7)
y>3

Note that this Boolean array preserves the 5x7 shape of `y`.

In [ ]:
(y>3).shape

But if we use it to index elements of `y`, the result has to be a 1D array:

In [ ]:
y[y>3]

Now suppose we have an array of data that has 4 columns, but we only want to keep the
first, second and fourth columns.  Here's a simple way to do that using
a Boolean array:

In [ ]:
X = np.arange(8).reshape((2,4)) + 1
print(X)
print(X[:,np.array([True,True,False,True])])

Now Boolean arrays and Boolean masking is a great feature but it will occasionally cause you grief if you forget that something is an array and that a Boolean test will often return an array of truth values instead of a truth value.  

Python will return an error if you try to use an array where a Boolean value is expected. The most common error of this sort is in an `if`-clause.  

In [ ]:
y = np.zeros((2,2))
# Hoping y is equivalent to False
if y:
    print('hi')

And the result of any Boolean test that can be interpreted as an elementwise operation will raise the same error:

In [ ]:
if y == 0:
    print('hi')

In [ ]:
y[0,1] = 2
y == 0

This is what Python means by saying "The truth value of an array with more than one element is ambiguous."  Generally it's a mixture of Truths and falsities, like real life.

## Data in the form of arrays

Many python modules that provide data do so in array form.  As an example,
we load the famous **iris data set** due to Ronald Fisher, familiar to many who've had a statistics class, but also to many who've a computer science or machine learning class in which data analysis plays a role.

In [ ]:
from sklearn.datasets import load_iris
data = load_iris()
features = data['data']
target = data['target']

The name `data` has been set to a dictionary.

In [ ]:
list(data.keys())

['data', 'target', 'target_names', 'DESCR', 'feature_names', 'filename']

In [ ]:
data

{'DESCR': '.. _iris_dataset:\n\nIris plants dataset\n--------------------\n\n**Data Set Characteristics:**\n\n    :Number of Instances: 150 (50 in each of three classes)\n    :Number of Attributes: 4 numeric, predictive attributes and the class\n    :Attribute Information:\n        - sepal length in cm\n        - sepal width in cm\n        - petal length in cm\n        - petal width in cm\n        - class:\n                - Iris-Setosa\n                - Iris-Versicolour\n                - Iris-Virginica\n                \n    :Summary Statistics:\n\n    ============== ==== ==== ======= ===== ====================\n                    Min  Max   Mean    SD   Class Correlation\n    ============== ==== ==== ======= ===== ====================\n    sepal length:   4.3  7.9   5.84   0.83    0.7826\n    sepal width:    2.0  4.4   3.05   0.43   -0.4194\n    petal length:   1.0  6.9   3.76   1.76    0.9490  (high!)\n    petal width:    0.1  2.5   1.20   0.76    0.9565  (high!)\n    ===========

The variable `features` has been set to `data['data']` and that is a `numpy` array:

In [ ]:
print(features.shape)
print(features.ndim)
print(features.dtype.name)
print(features.size)
print(type(features))
print(features[:10])

(150, 4)
2
float64
600
<class 'numpy.ndarray'>
[[5.1 3.5 1.4 0.2]
 [4.9 3.  1.4 0.2]
 [4.7 3.2 1.3 0.2]
 [4.6 3.1 1.5 0.2]
 [5.  3.6 1.4 0.2]
 [5.4 3.9 1.7 0.4]
 [4.6 3.4 1.4 0.3]
 [5.  3.4 1.5 0.2]
 [4.4 2.9 1.4 0.2]
 [4.9 3.1 1.5 0.1]]


The last thing printed is the first 10 rows of `features`.  This array represents data
about irises.  Each row represents a different iris and gives 4 measurements for that 
exemplar.  So there are 150 iris exemplars; with 4 measurements for each, that's 600 items of
data in the array (`features.size`).  The data is used for classification studies.

Let's apply what we've just learnt to pick a certain subset of the data. 
The first number
in each row represents the **sepal length** of that particular iris exemplar (to be demonstrated below).

Find the flowers whose sepal length is exactly 5.

In [ ]:
# 1D array: first column
first_col = features[:,0]
print(first_col == 5.0)

[False False False False  True False False  True False False False False
 False False False False False False False False False False False False
 False  True  True False False False False False False False False  True
 False False False False  True False False  True False False False False
 False  True False False False False False False False False False False
  True False False False False False False False False False False False
 False False False False False False False False False False False False
 False False False False False False False False False  True False False
 False False False False False False False False False False False False
 False False False False False False False False False False False False
 False False False False False False False False False False False False
 False False False False False False False False False False False False
 False False False False False False]


Then we use the Boolean array just constructed to index the rows, combined with a `:` to indicate we want all columns.

In [ ]:
features[first_col == 5.0,:]

array([[5. , 3.6, 1.4, 0.2],
       [5. , 3.4, 1.5, 0.2],
       [5. , 3. , 1.6, 0.2],
       [5. , 3.4, 1.6, 0.4],
       [5. , 3.2, 1.2, 0.2],
       [5. , 3.5, 1.3, 0.3],
       [5. , 3.5, 1.6, 0.6],
       [5. , 3.3, 1.4, 0.2],
       [5. , 2. , 3.5, 1. ],
       [5. , 2.3, 3.3, 1. ]])

So there are only 10 plants out of 150 that have sepal lengths of exactly 5.

Of course we can do this all in one step, with the same result:

In [ ]:
features[features[:,0] == 5.0,:]

array([[5. , 3.6, 1.4, 0.2],
       [5. , 3.4, 1.5, 0.2],
       [5. , 3. , 1.6, 0.2],
       [5. , 3.4, 1.6, 0.4],
       [5. , 3.2, 1.2, 0.2],
       [5. , 3.5, 1.3, 0.3],
       [5. , 3.5, 1.6, 0.6],
       [5. , 3.3, 1.4, 0.2],
       [5. , 2. , 3.5, 1. ],
       [5. , 2.3, 3.3, 1. ]])

Since we're indexing rows here, the `:` can also be left out:

In [ ]:
features[features[:,0] == 5.0]

array([[5. , 3.6, 1.4, 0.2],
       [5. , 3.4, 1.5, 0.2],
       [5. , 3. , 1.6, 0.2],
       [5. , 3.4, 1.6, 0.4],
       [5. , 3.2, 1.2, 0.2],
       [5. , 3.5, 1.3, 0.3],
       [5. , 3.5, 1.6, 0.6],
       [5. , 3.3, 1.4, 0.2],
       [5. , 2. , 3.5, 1. ],
       [5. , 2.3, 3.3, 1. ]])

A more interesting use of Boolean array indexing is to find all the irises of a particular class, using the variable `target` defined when we loaded the data; `target` is an array containing the class of each iris in the data set:

In [ ]:
print(type(target))
print(target.shape)
print(set(target))
# Grab the first rows of both arrays
print(features[0], target[0])
# Grab row indexed 90 from both arrays
print(features[90], target[90])

As the printouts indicate , `target` is a 1D array (a **vector**) of length 150, containing only the values 0,1, and 2.  These are the three classes to which an iris can belong.  There are exactly as many entries in the `target` array as there are rows in the `features` array. For any iris, its row-index in the `features` array is the index of its class in the `target`.  Above we printed the features and target class for irises 0 and 90.

We can take advantage of this structure to find all the irises of class 1 very efficiently.
Note that `target == 1` is a Boolean array of length 150.

In [ ]:
target==1

To find the irises of class 1, we simply use that Boolean array to specify the rows
we want in the `features` array:

In [ ]:
features[target==1,:].shape

There are 50 rows whose class is `1`; you can verify for yourselves that are also 50 rows for classes 2 and 3.  This is a balanced data set.

Going back to the dictionary we loaded when we loaded the datset, we can
find the meanings of our 4 columns:

In [ ]:
data['feature_names']

So our 4 columns have numerical data on these attributes. The column indexed 0 contains the sepal length measurement for 1 flower in our sample.  The column indexed 1 contains sepal width measurements. And so on.

We know there are three classes (0, 1, and 2). We can look up their names under the `'target_names'` key of the dictionary we got when we loaded the data.  

In [ ]:
data['target_names']

So class 0 represent the iris species `setosa`, and so on.

Here's what the entire target array looks like:

In [ ]:
data['target']

Find the irises whose species is 1, and whose sepal length is greater than 5.

In [ ]:
# New smaller data table consisting of those flowers of species 1.
species1 = features[target == 1,:]
print(len(species1))
species1[species1[:,0]>=5.0,:]
print(len(species1[species1[:,0]>=5.0,:]))

In [ ]:
print(species1[:,0]>=5.0)
sum(species1[:,0]>=5.0)

Now suppose we wanted to just see the **sepal lengths** (column 0) of the target class 1 irises, and compare them to the sepal lengths of the target class 0 irises, to see if sepal lengths provided a good way of telling the two classes apart.  We could do:

In [ ]:
print(features[target==1,0])
print(features[target==0,0])

And though there is some overlap, we see that there is a tendency for the column-0
value be higher in target 1 than it is target class 0.

## Cosine

One way of measuring the similarity between two arrays `x` and `y`. is to take their cosine.  The name for this similarity measure is quite appropriate.  The cosine of `x` and `y` is a measure of the geometric cosine of the angle between them.  It is 1 when that angle is 0 (i.e., `x` and `y` are identical in direction), and -1 when the vectors point in opposite directions, and 0 when the vectors are orthogonal (the do not share a component in any direction).  It is computed by taking the dot product of the unit vectors pointing in the same direction as `x` and `y`.  To do this we divide `x` and `y` by their Euclidean length (`LA.norm` in the code below) and take the dot product of the results.

In [ ]:
import numpy as np
import numpy.linalg as LA
x = np.array([3,4])
y = np.array([4,3])

def find_unit_vector(x):
    return x/LA.norm(x,ord=2)

def cosine (x,y):
    return (find_unit_vector(x)).dot(find_unit_vector(y))

The maximum possible value cosine can have for two vectors is `1.0`.  The maximum possible value is achieved when you compare a vector `x` with itself `x`.  So nothing can be more similar to `x` than `x` is to itself.  Whew.  That's reassuring.

Check my claim about how similar things are top themselves.

In [ ]:
cosine(x,x)

1.0

Let's mutliply x by -1 to get a vector pointing in exactly the opposite direction from `x`, and take the cosine of the two vectors.

In [ ]:
cosine(x,-x)

-1.0

`x` and `y` are different but not **that** different. What value does `cos` assign to their similarity?

In [ ]:
cosine(x,y)

0.96

What does `find_unit_vector` really do?  It finds a vector pointing in exactly the same direction as `x` but having length 1.  How similar will `x` be to its unit vector?  To find that out for yourself, create a Code cell and execute:

```
cosine(x, find_unit_vector(x))
```

Before doing that, look at the definition of cosine again and
try to predict what the value will be.

Does a unit vector really have have length 1?

In [ ]:
LA.norm(find_unit_vector(x))

1.0

In [ ]:
((135 * 6) + 120 + 40)* 1.11

1076.7

## Why use arrays?

Arrays provide all kinds of indexing convenience for accessing data that
is genuinely **tabular**.  Tabular data is data that has good reason for being represented in
a table;  that is, there is good reason to think of the rows as a unit, and good reason
to think of the columns as units.  And therefore, there is good reason to think we might want
to apply some computational operations to columns and rows.  We'll explore those ideas a bit more
in a future assignment.  

All kinds of data is fundamentally tabular, as we'll see.  But although that's very important,
it's probably not the main reason why so many scientists, engineers, and data analysts use `numpy`
and other programming tools like `matlab` and `R` that provide tables, or **matrices** (the mathematical
name for a similar concept), as a fundamental data structure.  The main reason is computational efficiency.

Having a single data structure with predictable types and elementwise basic arithmetic operations makes
it possible to write very efficient ways of performing more complex operations.  The code snippet. below
demonstrates this by providing a timing comparison between two ways of computing the sum of the
squares between 0 and 999 inclusive; the first is used in computing the variable `normal_py_sec`, is the ordinary Python way of computing such a sum:  Cook up a container of theose integers, square each of them and sum the result.
The second, stored in the varibale `np_sec`, uses numpy arrays.  There is actually a very important
mathematical operation **dot product** that multiplies all the corresponding elements of two vectors (elemntwise) and
sums the result.  Using this `precompiled` fast math operation we speed the code up by a factor of about
 80 on my machine, and your results, on your various home machines will be different, but in the same ballpark.
 

In [ ]:
import timeit
import numpy as np

py_sec = timeit.timeit('sum(x*x for x in range(1000))',
                              number=10000)

np_sec = timeit.timeit('na.dot(na)',
                            setup="import numpy as np; na=np.arange(1000)",
                            number=10000)

print(f"Normal Python: {py_sec} sec")

print(f"NumPy: {np_sec} sec")

Normal Python: 0.8127167950005969 sec
NumPy: 0.01427987700299127 sec


Note that numpy arrays in and of themselves don't help.  In the code snippet below, we multiply the `na` array times itself and sum the result.  This code is actually *slower* than the basic Python code, because it pays the overhead of creating the `numpy` array but doesn't exploit it by using the proper optimized operation.  The lesson here is that coding so as to  take maximum advantage of the efficiencies of `numpy` can be tricky, but if you have a good reason to be efficient, it's definitely worth some experimentation. That starts with knowing about the kinds of builtin array specific functions `numpy` offers.

In [ ]:
naive_np_sec = timeit.timeit('sum(na*na)',
                             setup="import numpy as np; na=np.arange(1000)",
                             number=10000)
print(("Naive NumPy: %f sec" % naive_np_sec))

Naive NumPy: 1.982286 sec
